README.md
# Cardano Verify Datasignature

<p align="left">

<img alt="Release" src="https://github.com/cardano-foundation/cardano-verify-datasignature/actions/workflows/release.yml/badge.svg?branch=main" />
<img alt="semantic-release: angular" src="https://img.shields.io/badge/semantic--release-angular-e10079?logo=semantic-release" />

</p>

A lightweight typescript library to verify a cip30 datasignature.

## 🚀 Getting Started

```zsh
npm i @cardano-foundation/cardano-verify-datasignature
```

## 🧐 Verification

You need a key and a signature from a [cip30 datasignature](https://cips.cardano.org/cips/cip30/#apisigndataaddraddresspayloadbytespromisedatasignature).

### How to get a cip30 datasignature?

There are multiple ways to create a cip30 data signature:

1. Make sure you have a cip30 compatible wallet installed (Nami, NuFi, Typhon Wallet, Flint, Gerowallet, Yoroi, ...).

2. Option A: Open your browser console and copy&paste this function:

```js
const signMessage = async (message, walletname) => {
  const api = await window.cardano[walletname].enable();
  const hexAddresses = await api.getRewardAddresses();
  const hexAddress = hexAddresses[0];
  let hexMessage = '';

  for (var i = 0, l = message.length; i < l; i++) {
    hexMessage += message.charCodeAt(i).toString(16);
  }

  try {
    const { signature, key } = await api.signData(hexAddress, hexMessage);
    console.log(signature, key);
  } catch (error) {
    console.warn(error);
  }
};
```

Usage example:

```js
signMessage('Hello World', 'yoroi').then((dataSignature) =>
  console.log(dataSignature)
);
```

3. Option B: Use the [signMessage](https://github.com/cardano-foundation/cardano-connect-with-wallet/blob/main/src/hooks/useCardano.ts#L133) function of the [cardano-connect-with-wallet](https://github.com/cardano-foundation/cardano-connect-with-wallet) library to get a valid key and signature.

4. Option C: If you want to use the cli to sign or verify data please checkout the [cardano-signer](https://github.com/gitmachtl/cardano-signer) by [gitmachtl](https://github.com/gitmachtl).

### What does it mean "to verify a signature"?

This function uses the public key (COSE_KEY) and checks if its corresponding private key has been used to sign the payload (data/message) within the signature (COSE_Sign1).

Furthermore an optional plain text message can be provided to check if both the plain text and the signed message are equal.

Another optional argument is a readable (bech32) address starting with (addr1/addr_test1/stake1/stake_test1) to test if this address belongs to the key that was used to sign the message.

```ts
const verifyDataSignature = require('@cardano-foundation/cardano-verify-datasignature');

const key =
  'a4010103272006215820b89526fd6bf4ba737c55ea90670d16a27f8de6cc1982349b3b676705a2f420c6';
const signature =
  '84582aa201276761646472657373581de118987c1612069d4080a0eb247820cb987fea81bddeaafdd41f996281a166686173686564f458264175677573746120416461204b696e672c20436f756e74657373206f66204c6f76656c61636558401712458b19f606b322982f6290c78529a235b56c0f1cec4f24b12a8660b40cd37f4c5440a465754089c462ed4b0d613bffaee3d1833516569fda4852f42a4a0f';
const message = 'Augusta Ada King, Countess of Lovelace';
const address = 'stake1uyvfslqkzgrf6syq5r4jg7pqewv8l65phh024lw5r7vk9qgznhyty';

console.log(verifyDataSignature(signature, key)); // true
console.log(verifyDataSignature(signature, key, message)); // true
console.log(verifyDataSignature(signature, key, message, address)); // true
console.log(
  verifyDataSignature(
    signature,
    key,
    message,
    'stake1_test1hweafkafrwf9ets85rs9gtk9qgzegwtg'
  )
); // false
console.log(
  verifyDataSignature(signature, key, 'Augusta Ada King, Countess of Lovelace!')
); // false
```


CHANGELOG.md
# Changelog

## [1.0.11](https://github.com/cardano-foundation/cardano-verify-datasignature/compare/v1.0.10...v1.0.11) (2022-12-21)


### Bug Fixes

* fill null payload with message parameter if provided. solves [#11](https://github.com/cardano-foundation/cardano-verify-datasignature/issues/11) ([71b1ae4](https://github.com/cardano-foundation/cardano-verify-datasignature/commit/71b1ae44315a8dbfecefcf6a4cdbeac8610f719c))

## [1.0.10](https://github.com/cardano-foundation/cardano-verify-datasignature/compare/v1.0.9...v1.0.10) (2022-12-20)


### Features

* add cardano-signer as a cli option to the readme ([5364349](https://github.com/cardano-foundation/cardano-verify-datasignature/commit/53643497d5dab0abc1e3b05a404e0d753d1730fe))

## [1.0.9](https://github.com/cardano-foundation/cardano-verify-datasignature/compare/v1.0.8...v1.0.9) (2022-12-20)


### Features

* add support for enterprise addresses should fix [#9](https://github.com/cardano-foundation/cardano-verify-datasignature/issues/9) ([619a2ce](https://github.com/cardano-foundation/cardano-verify-datasignature/commit/619a2ce8197fd6bc3f54629505adc22f401d7c9a))

## [1.0.8](https://github.com/cardano-foundation/cardano-verify-datasignature/compare/v1.0.7...v1.0.8) (2022-12-19)


### Features

* add more testcases to check if all cases of [#9](https://github.com/cardano-foundation/cardano-verify-datasignature/issues/9) and [#11](https://github.com/cardano-foundation/cardano-verify-datasignature/issues/11) are now handled correctly ([510e10d](https://github.com/cardano-foundation/cardano-verify-datasignature/commit/510e10dac49f155a65286ba5b48a90ee5257ebf4))

## [1.0.7](https://github.com/cardano-foundation/cardano-verify-datasignature/compare/v1.0.6...v1.0.7) (2022-12-18)


### Bug Fixes

* repair payment address verification and check for null payload ([e43225e](https://github.com/cardano-foundation/cardano-verify-datasignature/commit/e43225ea20901e55b68ef771a68b6d5b2258a9d9))

## [1.0.6](https://github.com/cardano-foundation/cardano-verify-datasignature/compare/v1.0.5...v1.0.6) (2022-12-18)


### Bug Fixes

* update @stricahq/cip08 version to handle hashed payloads (regarding to [#9](https://github.com/cardano-foundation/cardano-verify-datasignature/issues/9)) ([b6adf0e](https://github.com/cardano-foundation/cardano-verify-datasignature/commit/b6adf0ef5f0b12a083e0f819952d19b1ea14985c))

## [1.0.5](https://github.com/cardano-foundation/cardano-verify-datasignature/compare/v1.0.4...v1.0.5) (2022-12-13)


### Bug Fixes

* link types and preserve docstrings ([2a4eb6f](https://github.com/cardano-foundation/cardano-verify-datasignature/commit/2a4eb6fa340748b7021b9850d3d10991767a697a))

## [1.0.4](https://github.com/cardano-foundation/cardano-verify-datasignature/compare/v1.0.3...v1.0.4) (2022-12-13)


### Features

* add payment address verification ([e2e5341](https://github.com/cardano-foundation/cardano-verify-datasignature/commit/e2e5341aa382a6ac4e0fd98566904c10599fc968))

## [1.0.3](https://github.com/cardano-foundation/cardano-verify-datasignature/compare/v1.0.2...v1.0.3) (2022-12-12)


### Bug Fixes

* resolve bundle error unknown this ([b17be8f](https://github.com/cardano-foundation/cardano-verify-datasignature/commit/b17be8fd78b5a16d2a17e58650ffd75dd85975f1))

## [1.0.2](https://github.com/cardano-foundation/cardano-verify-datasignature/compare/v1.0.1...v1.0.2) (2022-12-12)


### Features

* add webpack to build the module for browser and commonjs ([9bf7bf4](https://github.com/cardano-foundation/cardano-verify-datasignature/commit/9bf7bf46453b3d1ee777517b5934a9551136b89d))

## [1.0.1](https://github.com/cardano-foundation/cardano-verify-datasignature/compare/v1.0.0...v1.0.1) (2022-12-12)


### Bug Fixes

* set the package name to the github repo name ([43e9842](https://github.com/cardano-foundation/cardano-verify-datasignature/commit/43e9842ba60d8fa074f87dee656910e1d2624e66))

## 1.0.0 (2022-12-12)


### Features

* add build pipeline ([7d6e954](https://github.com/cardano-foundation/cardano-verify-datasignature/commit/7d6e95470e4b86046c24c44b68a946983cc5ea4d))
* implement signature verification ([9f3820c](https://github.com/cardano-foundation/cardano-verify-datasignature/commit/9f3820c3d1184d6912aaf573431f668c15be4873))


index.test.ts
import verifySignature from './index';
import { blake2bHex } from 'blakejs';

describe('testing signature verification', () => {
  // https://cips.cardano.org/cips/cip30/#datasignature
  test('signature verification using cip30 datasignature', () => {
    const key =
      'a4010103272006215820b89526fd6bf4ba737c55ea90670d16a27f8de6cc1982349b3b676705a2f420c6';
    const signature =
      '84582aa201276761646472657373581de118987c1612069d4080a0eb247820cb987fea81bddeaafdd41f996281a166686173686564f458264175677573746120416461204b696e672c20436f756e74657373206f66204c6f76656c61636558401712458b19f606b322982f6290c78529a235b56c0f1cec4f24b12a8660b40cd37f4c5440a465754089c462ed4b0d613bffaee3d1833516569fda4852f42a4a0f';

    expect(verifySignature(signature, key)).toBe(true);
  });

  test('signature verification using cip30 datasignature. Additionally check the raw message within the signature', () => {
    const key =
      'a4010103272006215820b89526fd6bf4ba737c55ea90670d16a27f8de6cc1982349b3b676705a2f420c6';
    const signature =
      '84582aa201276761646472657373581de118987c1612069d4080a0eb247820cb987fea81bddeaafdd41f996281a166686173686564f458264175677573746120416461204b696e672c20436f756e74657373206f66204c6f76656c61636558401712458b19f606b322982f6290c78529a235b56c0f1cec4f24b12a8660b40cd37f4c5440a465754089c462ed4b0d613bffaee3d1833516569fda4852f42a4a0f';
    const message = 'Augusta Ada King, Countess of Lovelace';

    expect(verifySignature(signature, key, message)).toBe(true);
  });

  test('signature verification using cip30 datasignature with message check and readable address check', () => {
    const address =
      'stake1uyvfslqkzgrf6syq5r4jg7pqewv8l65phh024lw5r7vk9qgznhyty';
    const key =
      'a4010103272006215820b89526fd6bf4ba737c55ea90670d16a27f8de6cc1982349b3b676705a2f420c6';
    const signature =
      '84582aa201276761646472657373581de118987c1612069d4080a0eb247820cb987fea81bddeaafdd41f996281a166686173686564f458264175677573746120416461204b696e672c20436f756e74657373206f66204c6f76656c61636558401712458b19f606b322982f6290c78529a235b56c0f1cec4f24b12a8660b40cd37f4c5440a465754089c462ed4b0d613bffaee3d1833516569fda4852f42a4a0f';
    const message = '!Augusta Ada King, Countess of Lovelace!';

    expect(verifySignature(signature, key, message, address)).toBe(false);
  });

  test('signature verification using cip30 datasignature with message check should fail', () => {
    const key =
      'a4010103272006215820b89526fd6bf4ba737c55ea90670d16a27f8de6cc1982349b3b676705a2f420c6';
    const signature =
      '84582aa201276761646472657373581de118987c1612069d4080a0eb247820cb987fea81bddeaafdd41f996281a166686173686564f458264175677573746120416461204b696e672c20436f756e74657373206f66204c6f76656c61636558401712458b19f606b322982f6290c78529a235b56c0f1cec4f24b12a8660b40cd37f4c5440a465754089c462ed4b0d613bffaee3d1833516569fda4852f42a4a0f';
    const message = '!Augusta Ada King, Countess of Lovelace!';

    expect(verifySignature(signature, key, message)).toBe(false);
  });

  test('signature verification using cip30 datasignature with address check should fail', () => {
    const address =
      'stake_test1uzmggtulkyt5df9rpmqvh9acuc4etr5vntehx65nq2uz2mg8293u0';
    const key =
      'a4010103272006215820b89526fd6bf4ba737c55ea90670d16a27f8de6cc1982349b3b676705a2f420c6';
    const signature =
      '84582aa201276761646472657373581de118987c1612069d4080a0eb247820cb987fea81bddeaafdd41f996281a166686173686564f458264175677573746120416461204b696e672c20436f756e74657373206f66204c6f76656c61636558401712458b19f606b322982f6290c78529a235b56c0f1cec4f24b12a8660b40cd37f4c5440a465754089c462ed4b0d613bffaee3d1833516569fda4852f42a4a0f';

    expect(verifySignature(signature, key, undefined, address)).toBe(false);
  });

  test('signature verification with payment address', () => {
    const address =
      'addr1qxtu4w2rq2mdguw4fkms2ge4m070nq8cmlyjfhghwlh8sjscnp7pvysxn4qgpg8ty3uzpjuc0l4gr0w74t7ag8uev2qseuyw6u';
    const key =
      'a4010103272006215820b89526fd6bf4ba737c55ea90670d16a27f8de6cc1982349b3b676705a2f420c6';
    const signature =
      '84582aa201276761646472657373581de118987c1612069d4080a0eb247820cb987fea81bddeaafdd41f996281a166686173686564f458264175677573746120416461204b696e672c20436f756e74657373206f66204c6f76656c61636558401712458b19f606b322982f6290c78529a235b56c0f1cec4f24b12a8660b40cd37f4c5440a465754089c462ed4b0d613bffaee3d1833516569fda4852f42a4a0f';

    expect(verifySignature(signature, key, undefined, address)).toBe(true);
  });

  test('hashed payload with stake address that should be valid and return true', () => {
    const signature =
      '84582aa201276761646472657373581de0c13582aec9a44fcc6d984be003c5058c660e1d2ff1370fd8b49ba73fa166686173686564f5581c40843181253eb1ff2258ab39c3463ec0edf5e713b73c5482c0ca798f5840a4cdec07ba8c1184aa74d1c3516fc6602a35d2db847510cf98c102653c15c7664f136314f920150a081870aef77ed49780ca58873bd5d62e744b968a89435906';
    const key =
      'a40101032720062158209be513df12b3fabe7c1b8c3f9fab0968eb2168d5689bf981c2f7c35b11718b27';

    expect(verifySignature(signature, key)).toBe(true);
  });

  test('signature signed with payment key should be valid', () => {
    const signature =
      '845846a20127676164647265737358390197cab94302b6d471d54db7052335dbfcf980f8dfc924dd1777ee784a18987c1612069d4080a0eb247820cb987fea81bddeaafdd41f996281a166686173686564f44b48656c6c6f20576f726c645840b65cb33e107a692605a479811a8405e44eeac5217c6ef92b79c221c2309305ec2db927fb75a7d197602e1eb2e663dae227aa7c0510b6484f5591b2b4bd47b70d';
    const key =
      'a4010103272006215820472be3f30b51ead6d020e0d370774861e242ca23eaca2f4eff4ddb8eaa3abefd';
    expect(verifySignature(signature, key)).toBe(true);
  });

  test('signature signed with payment key and a bech32 key for the verification should be valid', () => {
    const signature =
      '845846a20127676164647265737358390197cab94302b6d471d54db7052335dbfcf980f8dfc924dd1777ee784a18987c1612069d4080a0eb247820cb987fea81bddeaafdd41f996281a166686173686564f44b68656c6c6f20776f726c645840d1a116c02d6ea035928f85ae59b7181f2b1cc31e673e1d890a5f92d6d3def81d667664ac091b42bd746dcd60b1a735b6c0ecceaa6672434ecf719b380db87808';
    const key =
      'a4010103272006215820472be3f30b51ead6d020e0d370774861e242ca23eaca2f4eff4ddb8eaa3abefd';
    const address =
      'addr1qxtu4w2rq2mdguw4fkms2ge4m070nq8cmlyjfhghwlh8sjscnp7pvysxn4qgpg8ty3uzpjuc0l4gr0w74t7ag8uev2qseuyw6u';
    expect(verifySignature(signature, key, undefined, address)).toBe(true);
  });

  test('signature signed with payment key and a wrong bech32 key for the verification should be invalid', () => {
    const signature =
      '845846a20127676164647265737358390197cab94302b6d471d54db7052335dbfcf980f8dfc924dd1777ee784a18987c1612069d4080a0eb247820cb987fea81bddeaafdd41f996281a166686173686564f44b48656c6c6f20576f726c645840b65cb33e107a692605a479811a8405e44eeac5217c6ef92b79c221c2309305ec2db927fb75a7d197602e1eb2e663dae227aa7c0510b6484f5591b2b4bd47b70d';
    const key =
      'a4010103272006215820472be3f30b51ead6d020e0d370774861e242ca23eaca2f4eff4ddb8eaa3abefd';
    const address =
      'addr1qykeu08ymqftykmddg8fuc2d78huz8aa48afk00twn7ye65zq5rvkr89ftn4tvj39dk0xxzk6un9apujewr2lj2wppeqxsygwc';
    expect(verifySignature(signature, key, address)).toBe(false);
  });

  test('check signature with null payload should be false', () => {
    const signature =
      '84582aa201276761646472657373581de0c13582aec9a44fcc6d984be003c5058c660e1d2ff1370fd8b49ba73fa166686173686564f4f658400a0dd23e867292a4c2eb692f63016e3f61294686f672065fcc377f665cff6b25c430619060b536073cfd2355ab6c6bcec9d7ecbfb588f7b0aa5967f1b8559300';
    const key =
      'a40101032720062158209be513df12b3fabe7c1b8c3f9fab0968eb2168d5689bf981c2f7c35b11718b27';
    expect(verifySignature(signature, key)).toBe(false);
  });

  test('check signature with correct plain text message but empty payload should be true', () => {
    const signature =
      '84582aa201276761646472657373581de0c13582aec9a44fcc6d984be003c5058c660e1d2ff1370fd8b49ba73fa166686173686564f4f658400a0dd23e867292a4c2eb692f63016e3f61294686f672065fcc377f665cff6b25c430619060b536073cfd2355ab6c6bcec9d7ecbfb588f7b0aa5967f1b8559300';
    const key =
      'a40101032720062158209be513df12b3fabe7c1b8c3f9fab0968eb2168d5689bf981c2f7c35b11718b27';
    const message = 'Hello world';
    expect(verifySignature(signature, key, message)).toBe(true);
  });

  test('hashed payload with plain text message that should be valid and return true', () => {
    const signature =
      '84582aa201276761646472657373581de0c13582aec9a44fcc6d984be003c5058c660e1d2ff1370fd8b49ba73fa166686173686564f4f658400a0dd23e867292a4c2eb692f63016e3f61294686f672065fcc377f665cff6b25c430619060b536073cfd2355ab6c6bcec9d7ecbfb588f7b0aa5967f1b8559300';
    const key =
      'a40101032720062158209be513df12b3fabe7c1b8c3f9fab0968eb2168d5689bf981c2f7c35b11718b27';
    const message = 'Hello world';

    expect(verifySignature(signature, key, message)).toBe(true);
  });

  test('hashed payload with hashed text message that should be valid and return true', () => {
    const signature =
      '84582aa201276761646472657373581de0c13582aec9a44fcc6d984be003c5058c660e1d2ff1370fd8b49ba73fa166686173686564f5581c40843181253eb1ff2258ab39c3463ec0edf5e713b73c5482c0ca798f5840a4cdec07ba8c1184aa74d1c3516fc6602a35d2db847510cf98c102653c15c7664f136314f920150a081870aef77ed49780ca58873bd5d62e744b968a89435906';
    const key =
      'a40101032720062158209be513df12b3fabe7c1b8c3f9fab0968eb2168d5689bf981c2f7c35b11718b27';
    const message = blake2bHex('Hello world', undefined, 28); // 28 * 8 bit = 224 bit

    expect(verifySignature(signature, key, message)).toBe(true);
  });

  test('test verification with provided enterprise address', () => {
    const signature =
      '84582aa201276761646472657373581d617863b5c43bdf0a06608abc82f0573a549714ff69166074dcdde393d8a166686173686564f44b48656c6c6f20776f726c645840fc58155f0cee05bc00e7299af1df1f159ac82a46a055786b259657934eff346eec81349d4678ceabc79f213c66a2bdbfd4ea5d9ebdc630bee5ac9cce75cfc001';
    const key =
      'a4010103272006215820755b017578b701dc9ddd4eaee67015b4ca8baf66293b7b1d204df426c0ceccb9';
    const message = 'Hello world';

    expect(
      verifySignature(
        signature,
        key,
        message,
        'addr1v9ux8dwy800s5pnq327g9uzh8f2fw98ldytxqaxumh3e8kqumfr6d'
      )
    ).toBe(true);
  });
});


webpack.config.js
const webpack = require('webpack');
const path = require('path');
const { CleanWebpackPlugin } = require('clean-webpack-plugin');

const config = {
  mode: 'production',
  entry: path.resolve(__dirname, 'index.ts'),
  output: {
    path: path.resolve(__dirname, 'dist'),
    filename: 'index.js',
    library: {
      name: 'CardanoVerifyDataSignature',
      type: 'var',
      export: 'default',
    },
    libraryTarget: 'umd',
    globalObject: 'this',
  },
  experiments: {
    asyncWebAssembly: true,
    syncWebAssembly: true,
  },
  devtool: 'source-map',
  module: {
    rules: [
      {
        test: /\.(js|mjs|jsx|ts|tsx)$/,
        exclude: /node_modules/,
        use: {
          loader: 'babel-loader',
          options: {
            presets: [
              '@babel/preset-env',
              ['@babel/preset-typescript', { allowNamespaces: true }],
            ],
          },
        },
      },
    ],
  },
  resolve: {
    extensions: ['*', '.js', '.jsx', '.ts', '.tsx', '.scss'],
    fallback: {
      stream: require.resolve('stream-browserify'),
      buffer: require.resolve('buffer/'),
    },
  },
  plugins: [
    new webpack.ProgressPlugin(),
    new CleanWebpackPlugin({
      verbose: true,
      cleanStaleWebpackAssets: true,
    }),
    new webpack.ProvidePlugin({
      Buffer: ['buffer', 'Buffer'],
    }),
  ],
  infrastructureLogging: {
    level: 'info',
  },
};

module.exports = config;


index.ts
import { Bip32PublicKey } from '@stricahq/bip32ed25519';
import { getPublicKeyFromCoseKey, CoseSign1 } from '@stricahq/cip08';
import { Decoder } from '@stricahq/cbors';
import {
  BaseAddress,
  EnterpriseAddress,
  RewardAddress,
} from '@stricahq/typhonjs/dist/address';
import { utils } from '@stricahq/typhonjs';
import { blake2bHex } from 'blakejs';

const Network = {
  MAINNET: 1,
  TESTNET: 0,
};

// This function is an adjusted version of
// https://github.com/StricaHQ/cip08/blob/39223edfe2c20c37a78e49aaab77761f9534cef1/src/coseSign1.ts#L32
const CoseSign1FromCborWithPayload = (cbor: string, payload: Buffer) => {
  const decoded = Decoder.decode(Buffer.from(cbor, 'hex'));

  if (!(decoded.value instanceof Array)) throw Error('Invalid CBOR');
  if (decoded.value.length !== 4) throw Error('Invalid COSE_SIGN1');

  let protectedMap;
  // Decode and Set ProtectedMap
  const protectedSerialized = decoded.value[0];
  try {
    protectedMap = Decoder.decode(protectedSerialized).value;
    if (!(protectedMap instanceof Map)) {
      throw Error();
    }
  } catch (error) {
    throw Error('Invalid protected');
  }

  // Set UnProtectedMap
  const unProtectedMap = decoded.value[1];
  if (!(unProtectedMap instanceof Map)) throw Error('Invalid unprotected');

  // Set Signature
  const signature = decoded.value[3];

  return new CoseSign1({
    protectedMap,
    unProtectedMap,
    payload,
    signature,
  });
};

const addSpanBytesToObject = (obj: any, span: [number, number]): any => {
  const spanObj = obj;
  spanObj.byteSpan = span;
  spanObj.getByteSpan = function (): [number, number] {
    return this.byteSpan;
  };

  return spanObj;
};

/**
 * This method can be used to verify a cip30 data signature.
 *
 * https://cips.cardano.org/cips/cip30/#datasignature
 * type DataSignature = {
 *   signature: cbor\<COSE_Sign1>,
 *   key: cbor\<COSE_Key>,
 * };
 *
 * Optional:
 *
 * Furthermore a plain text message can be provided to check if both
 * the plain text and the signed message are equal
 *
 * Also a payment (addr1/addr_test1) or stake address (stake1/stake_test1) can be provided
 * to verify that this readable address belongs to the signature
 *
 * @param {string} signature - A COSE_Sign1 cbor string.
 * @param {string} key - A COSE_Key cbor string
 * @param {string} message - An optional plain text message
 * @param {string} address - An optional address starting with (addr1/addr_test1/stake1/stake_test1)
 * @returns {string} boolean - True if the verification was successful otherwise false
 */
const verifySignature = (
  signature: string,
  key: string,
  message?: string,
  address?: string
): boolean => {
  const publicKeyBuffer = getPublicKeyFromCoseKey(key);
  let coseSign1 = CoseSign1.fromCbor(signature);

  if (message) {
    const decoded = Decoder.decode(Buffer.from(signature, 'hex'));
    const payload: Buffer = decoded.value[2];
    const unprotectedMap: Map<any, any> = decoded?.value[1];
    const isHashed =
      unprotectedMap && unprotectedMap.get('hashed')
        ? unprotectedMap.get('hashed')
        : false;

    if (
      payload === null ||
      typeof payload === 'undefined' ||
      (payload.toString() === '' && message !== '')
    ) {
      if (isHashed) {
        coseSign1 = CoseSign1FromCborWithPayload(
          signature,
          Buffer.from(blake2bHex(message, undefined, 28))
        );
      } else {
        coseSign1 = CoseSign1FromCborWithPayload(
          signature,
          Buffer.from(message)
        );
      }
    }

    if (isHashed && !/^[0-9a-fA-F]+$/.test(message)) {
      message = blake2bHex(message, undefined, 28); // 28 * 8 bit = 224 bit ~> blake2b224
    }

    if (isHashed && payload && payload.toString('hex') !== message) {
      return false;
    } else if (!isHashed && payload && payload.toString('utf8') !== message) {
      return false;
    }
  }

  if (address) {
    let providedAddress = address;
    let network = Network.MAINNET;
    const paymentAddress = utils.getAddressFromBech32(address) as BaseAddress;
    const coseSign1PublicKey = new Bip32PublicKey(publicKeyBuffer);

    const credential = {
      hash: coseSign1PublicKey.toPublicKey().hash().toString('hex'),
      type: 0,
    };

    if (address.startsWith('addr')) {
      if (address.startsWith('addr_test1')) {
        network = Network.TESTNET;
      }

      if (paymentAddress.stakeCredential) {
        const paymentAddressBech32 = new BaseAddress(
          network,
          credential,
          paymentAddress.stakeCredential
        ).getBech32();

        if (address !== paymentAddressBech32) {
          // Test whether the key is a stake key to which this payment key belongs
          const extractedRewardAddress = new RewardAddress(
            network,
            paymentAddress.stakeCredential
          ).getBech32();

          const rewardAddress = new RewardAddress(
            network,
            credential
          ).getBech32();

          if (rewardAddress !== extractedRewardAddress) {
            return false;
          }
        }
      } else {
        // An EnterpriseAddress does not have a stake credential
        const enterpriseAddress = new EnterpriseAddress(
          network,
          credential
        ).getBech32();
        if (enterpriseAddress !== providedAddress) {
          return false;
        }
      }
    } else if (address.startsWith('stake')) {
      if (address.startsWith('stake_test1')) {
        network = Network.TESTNET;
      }

      const rewardAddress = new RewardAddress(network, credential).getBech32();

      if (rewardAddress !== providedAddress) {
        return false;
      }
    } else {
      return false;
    }
  }

  return coseSign1.verifySignature({
    publicKeyBuffer: publicKeyBuffer,
  });
};

export default verifySignature;
